File: Lab_1/Lab_1.7/Bill.cs
namespace Lab_1._7;

public class Bill
{
    public string LastName { get; private set; }
    public string PhoneNumber { get; private set; }
    public double MinuteRate { get; private set; }
    public double Discount { get; private set; }
    public Time StartTime { get; private set; }
    public Time EndTime { get; private set; }
    public double TotalAmount { get; private set; }

    public bool Init(string lastName, string phoneNumber, double minuteRate, double discount, Time startTime, Time endTime)
    {
        if (string.IsNullOrEmpty(lastName) || string.IsNullOrEmpty(phoneNumber) || minuteRate <= 0 || discount < 0 || discount > 100 || startTime == null || endTime == null)
        {
            Console.WriteLine("Invalid input! Please check your data.");
            return false;
        }

        LastName = lastName;
        PhoneNumber = phoneNumber;
        MinuteRate = minuteRate;
        Discount = discount;
        StartTime = startTime;
        EndTime = endTime;

        CalculateTotalAmount();
        return true;
    }

    public void Read()
    {
        bool initialized = false;
        do
        {
            Console.WriteLine("Enter details for bill:");
            Console.Write("Last Name: ");
            string lastName = Console.ReadLine();
            Console.Write("Phone Number: ");
            string phoneNumber = Console.ReadLine();
            Console.Write("Minute Rate: ");
            double minuteRate;
            double.TryParse(Console.ReadLine(), out minuteRate);
            Console.Write("Discount: ");
            double discount;
            double.TryParse(Console.ReadLine(), out discount);
            bool ifRepeat = false;
            Time startTime = new();
            Time endTime = new();
            do
            {
                Console.WriteLine("Enter start time: ");
                startTime.Read();
                Console.WriteLine("Enter end time: ");
                endTime.Read();
                /* if (!startTime.IsEqualTo(endTime))
                 {
                     ifRepeat = true;
                     Console.WriteLine("\n\tThe end time cannot be equal to the start time");
                 }*/
            } while (ifRepeat);
            initialized = Init(lastName, phoneNumber, minuteRate, discount, startTime, endTime);
        } while (!initialized);
    }
    public void Display()
    {

        Console.WriteLine($"Last Name: {LastName}");
        Console.WriteLine($"Phone Number: {PhoneNumber}");
        Console.WriteLine($"Minute Rate: {MinuteRate}");
        Console.WriteLine($"Discount: {Discount}");
        Console.WriteLine("Start Time: " + StartTime.ToString());
        Console.WriteLine("End Time: " + EndTime.ToString());
        Console.WriteLine($"Total Amount: {TotalAmount} UAH");
    }
    private void CalculateTotalAmount()
    {
        double startTimeInMinutes = StartTime.Hour * 60 + StartTime.Minute + StartTime.Second / 60;
        double endTimeInMinutes = EndTime.Hour * 60 + EndTime.Minute + EndTime.Second / 60;

        if (endTimeInMinutes < startTimeInMinutes)
        {
            endTimeInMinutes += 24 * 60; 
        }

        double durationInMinutes = endTimeInMinutes - startTimeInMinutes;
        if (durationInMinutes % 1 != 0) { durationInMinutes = (int)durationInMinutes++; }
        double totalCost = durationInMinutes * MinuteRate;
        double discountAmount = totalCost * (Discount / 100);
        TotalAmount = totalCost - discountAmount;
    }

    public override string ToString() => $"Last Name: {LastName}, Phone Number: {PhoneNumber}, Minute Rate: {MinuteRate}, Discount: {Discount}%, StartTime: {StartTime.ToString()}, EndTime: {EndTime.ToString()}, Total Amount: {TotalAmount} UAH";
}

File: Lab_2/Lab_2.7_c++/Bill.cpp
#include <iostream>
#include <string>
#include <cmath>
#include <stdlib.h>
#include <sstream>
#include <iomanip>
#include "Bill.h"
using namespace std;

Bill::Bill() : minuteRate(0), discount(0), totalAmount(0) { }
Bill::Bill(const std::string& lastName, const std::string& phoneNumber, double minuteRate, double discount, const Time& startTime, const Time& endTime) {
	this->lastName = lastName;
	this->phoneNumber = phoneNumber;
	this->minuteRate = minuteRate;
	this->discount = discount;
	this->startTime = startTime;
	this->endTime = endTime;
	CalculateTotalAmount();
}
Bill::Bill(const Bill& other) {
	lastName = other.lastName;
	phoneNumber = other.phoneNumber;
	minuteRate = other.minuteRate;
	discount = other.discount;
	startTime = other.startTime;
	endTime = other.endTime;
	totalAmount = other.totalAmount;
	CalculateTotalAmount();
}

void Bill::CalculateTotalAmount() {
	double startTimeInMinutes = startTime.GetHour() * 60 + startTime.GetMinute() + startTime.GetSecond() / 60;
	double endTimeInMinutes = endTime.GetHour() * 60 + endTime.GetMinute() + endTime.GetSecond() / 60;

	if (endTimeInMinutes < startTimeInMinutes) {
		endTimeInMinutes += 24 * 60;
	}

	int durationInMinutes = endTimeInMinutes - startTimeInMinutes;
	if (durationInMinutes % 1 != 0) { durationInMinutes = (durationInMinutes++); }
	double totalCost = durationInMinutes * minuteRate;
	double discountAmount = totalCost * (discount / 100);
	totalAmount = totalCost - discountAmount;
}

Bill& Bill::operator =(const Bill& t)
{
	lastName = t.lastName;
	phoneNumber = t.phoneNumber;
	minuteRate = t.minuteRate;
	discount = t.discount;
	startTime = t.startTime;
	endTime = t.endTime;
	totalAmount = t.totalAmount;
	return *this;
}
std::ostream& operator<<(std::ostream& out, const Bill& bill) {
	out << "Last Name: " << bill.lastName << std::endl;
	out << "Phone Number: " << bill.phoneNumber << std::endl;
	out << "Minute Rate: " << bill.minuteRate << std::endl;
	out << "Discount: " << bill.discount << std::endl;
	out << "Start Time: " << bill.startTime << std::endl;
	out << "End Time: " << bill.endTime << std::endl;
	out << "Total Amount: " << bill.totalAmount << " UAH" << std::endl;
	return out;
}

std::istream& operator>>(std::istream& in, Bill& bill) {
	std::cout << "Enter details for bill:" << std::endl;
	std::cout << "Last Name: ";
	in >> bill.lastName;
	std::cout << "Phone Number: ";
	in >> bill.phoneNumber;
	std::cout << "Minute Rate: ";
	in >> bill.minuteRate;
	std::cout << "Discount: ";
	in >> bill.discount;

	std::cout << "Enter start time:" << std::endl;
	in >> bill.startTime;
	std::cout << "Enter end time:" << std::endl;
	in >> bill.endTime;
	return in;
}
Bill::operator string() const
{
	stringstream ss;
	ss << "Last Name: " << lastName << std::endl;
	ss << "Phone Number: " << phoneNumber << std::endl;
	ss << "Minute Rate: " << minuteRate << std::endl;
	ss << "Discount: " << discount << std::endl;
	ss << "Start Time: " << startTime << std::endl;
	ss << "End Time: " << endTime << std::endl;
	ss << "Total Amount: " << totalAmount << " UAH" << std::endl;
	return ss.str();
}

Bill& Bill::operator ++()
{
	++minuteRate;
	return *this;
}
Bill& Bill::operator --()
{
	--minuteRate;
	return *this;
}
Bill Bill::operator ++(int)
{
	Bill tmp = *this;
	++discount;
	return tmp;
}
Bill Bill::operator --(int)
{
	Bill tmp = *this;
	--discount;
	return tmp;
}
